pkg/kubelet/cm/memorymanager/memory_manager.go

```diff
@@ -0,0 +1,408 @@
+	AddContainer(p *v1.Pod, c *v1.Container, containerID string) error
+
+	// Allocate is called to pre-allocate memory resources during Pod admission.
+	// This must be called at some point prior to the AddContainer() call for a container, e.g. at pod admission time.
+	Allocate(pod *v1.Pod, container *v1.Container) error
+
+	// RemoveContainer is called after Kubelet decides to kill or delete a
+	// container. After this call, the memory manager stops trying to reconcile
+	// that container, and any memory allocated to the container are freed.
+	RemoveContainer(containerID string) error
+
+	// State returns a read-only interface to the internal memory manager state.
+	State() state.Reader
+
+	// GetTopologyHints implements the topologymanager.HintProvider Interface
+	// and is consulted to achieve NUMA aware resource alignment among this
+	// and other resource controllers.
+	GetTopologyHints(*v1.Pod, *v1.Container) map[string][]topologymanager.TopologyHint
+
+	// GetPodTopologyHints implements the topologymanager.HintProvider Interface
+	// and is consulted to achieve NUMA aware resource alignment among this
+	// and other resource controllers.
+	GetPodTopologyHints(*v1.Pod) map[string][]topologymanager.TopologyHint
+}
+
+type manager struct {
+	sync.Mutex
+	policy Policy
+
+	// state allows to restore information regarding memory allocation for guaranteed pods
+	// in the case of the kubelet restart
+	state state.State
+
+	// containerRuntime is the container runtime service interface needed
+	// to make UpdateContainerResources() calls against the containers.
+	containerRuntime runtimeService
+
+	// activePods is a method for listing active pods on the node
+	// so all the containers can be updated in the reconciliation loop.
+	activePods ActivePodsFunc
+
+	// podStatusProvider provides a method for obtaining pod statuses
+	// and the containerID of their containers
+	podStatusProvider status.PodStatusProvider
+
+	// containerMap provides a mapping from (pod, container) -> containerID
+	// for all containers a pod
+	containerMap containermap.ContainerMap
+
+	nodeAllocatableReservation v1.ResourceList
+
+	// sourcesReady provides the readiness of kubelet configuration sources such as apiserver update readiness.
+	// We use it to determine when we can purge inactive pods from checkpointed state.
+	sourcesReady config.SourcesReady
+
+	// stateFileDirectory holds the directory where the state file for checkpoints is held.
+	stateFileDirectory string
+}
+
+var _ Manager = &manager{}
+
+// NewManager returns new instance of the memory manager
+func NewManager(policyName string, machineInfo *cadvisorapi.MachineInfo, nodeAllocatableReservation v1.ResourceList, reservedMemory map[int]map[v1.ResourceName]resource.Quantity, stateFileDirectory string, affinity topologymanager.Store) (Manager, error) {
+	var policy Policy
+
+	switch policyType(policyName) {
+
+	case policyTypeNone:
+		policy = NewPolicyNone()
+
+	case policyTypeStatic:
```

Comment:
we are mixing terminology here.

i thought static was removed above for singleNUMA?

Reply:
We agreed to use a static policy, similar to the CPU manager because it does not really provide single NUMA node allocation(when the container memory requests bigger than the amount of the allocatable memory on the single NUMA node), the KEP also is using static policy, see - https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1769-memory-manager#memory-manager-policy-flag